// @algorandfoundation/algokit-utils@6.2.1/transaction/perform-atomic-transaction-composer-simulate.mjs
import algosdk from 'algosdk';

var modelsv2 = algosdk.modelsv2;
/**
 * Performs a simulation of the transactions loaded into the given AtomicTransactionComposer.
 * @param atc The AtomicTransactionComposer with transaction(s) loaded.
 * @param algod An Algod client to perform the simulation.
 * @returns The simulation result, which includes various details about how the transactions would be processed.
 */
async function performAtomicTransactionComposerSimulate(atc, algod) {
    const unsignedTransactionsSigners = atc.buildGroup();
    const decodedSignedTransactions = unsignedTransactionsSigners.map((ts) => algosdk.encodeUnsignedSimulateTransaction(ts.txn));
    const simulateRequest = new modelsv2.SimulateRequest({
        allowEmptySignatures: true,
        allowMoreLogging: true,
        execTraceConfig: new modelsv2.SimulateTraceConfig({
            enable: true,
            scratchChange: true,
            stackChange: true,
            stateChange: true,
        }),
        txnGroups: [
            new modelsv2.SimulateRequestTransactionGroup({
                txns: decodedSignedTransactions.map((txn) => algosdk.decodeObj(txn)),
            }),
        ],
    });
    const simulateResult = await algod.simulateTransactions(simulateRequest).do();
    return simulateResult;
}

export { performAtomicTransactionComposerSimulate };
