// @algorandfoundation/algokit-utils@6.2.1/types/client-manager.mjs
import algosdk from 'algosdk';
import { AlgoHttpClientWithRetry } from './algo-http-client-with-retry.mjs';
import { ApplicationClient } from './app-client.mjs';
import { TestNetDispenserApiClient } from './dispenser-client.mjs';

var Kmd = algosdk.Kmd;
var Indexer = algosdk.Indexer;
var IntDecoding = algosdk.IntDecoding;
/** Exposes access to various API clients. */
class ClientManager {
    /**
     * algosdk clients or config for interacting with the official Algorand APIs.
     * @param clientsOrConfig The clients or config to use
     * @example Algod client only
     * ```typescript
     * const clientManager = new ClientManager({ algod: algodClient })
     * ```
     * @example All clients
     * ```typescript
     * const clientManager = new ClientManager({ algod: algodClient, indexer: indexerClient, kmd: kmdClient })
     * ```
     * @example Algod config only
     * ```typescript
     * const clientManager = new ClientManager({ algodConfig })
     * ```
     * @example All client configs
     * ```typescript
     * const clientManager = new ClientManager({ algodConfig, indexerConfig, kmdConfig })
     * ```
     */
    constructor(clientsOrConfig) {
        const _clients = 'algod' in clientsOrConfig
            ? clientsOrConfig
            : {
                algod: ClientManager.getAlgodClient(clientsOrConfig.algodConfig),
                indexer: clientsOrConfig.indexerConfig ? ClientManager.getIndexerClient(clientsOrConfig.indexerConfig) : undefined,
                kmd: clientsOrConfig.kmdConfig ? ClientManager.getKmdClient(clientsOrConfig.kmdConfig) : undefined,
            };
        this._algod = _clients.algod;
        this._indexer = _clients.indexer;
        this._kmd = _clients.kmd;
    }
    /** Returns an algosdk Algod API client. */
    get algod() {
        return this._algod;
    }
    /** Returns an algosdk Indexer API client or throws an error if it's not been provided. */
    get indexer() {
        if (!this._indexer)
            throw new Error('Attempt to use Indexer client in AlgoKit instance with no Indexer configured');
        return this._indexer;
    }
    /** Returns an algosdk KMD API client or throws an error if it's not been provided. */
    get kmd() {
        if (!this._kmd)
            throw new Error('Attempt to use Kmd client in AlgoKit instance with no Kmd configured');
        return this._kmd;
    }
    /**
     * Get details about the current network.
     * @example Getting genesis ID
     * ```typescript
     * const network = await networkClient.network()
     * const genesisId = network.genesisId
     * ```
     * @returns The current network details
     */
    async network() {
        if (!this._getNetworkPromise) {
            this._getNetworkPromise = this._algod.getTransactionParams().do();
        }
        const params = await this._getNetworkPromise;
        return {
            isTestNet: ['testnet-v1.0', 'testnet-v1', 'testnet'].includes(params.genesisID),
            isMainNet: ['mainnet-v1.0', 'mainnet-v1', 'mainnet'].includes(params.genesisID),
            isLocalNet: ClientManager.genesisIdIsLocalNet(params.genesisID),
            genesisId: params.genesisID,
            genesisHash: params.genesisHash,
        };
    }
    /**
     * Returns true if the given network genesisId is associated with a LocalNet network.
     * @param genesisId The network genesis ID
     * @returns Whether the given genesis ID is associated with a LocalNet network
     */
    static genesisIdIsLocalNet(genesisId) {
        return genesisId === 'devnet-v1' || genesisId === 'sandnet-v1' || genesisId === 'dockernet-v1';
    }
    /**
     * Returns true if the current network is LocalNet.
     * @returns True if the current network is LocalNet.
     */
    async isLocalNet() {
        return (await this.network()).isLocalNet;
    }
    /**
     * Returns true if the current network is TestNet.
     * @returns True if the current network is TestNet.
     */
    async isTestNet() {
        return (await this.network()).isTestNet;
    }
    /**
     * Returns true if the current network is MainNet.
     * @returns True if the current network is MainNet.
     */
    async isMainNet() {
        return (await this.network()).isMainNet;
    }
    /**
     * Returns a TestNet Dispenser API client.
     * Refer to [docs](https://github.com/algorandfoundation/algokit/blob/main/docs/testnet_api.md) on guidance to obtain an access token.
     *
     * @param params An object containing parameters for the TestNetDispenserApiClient class.
     *  Or null if you want the client to load the access token from the environment variable `ALGOKIT_DISPENSER_ACCESS_TOKEN`.
     * @example
     * const client = clientManager.getTestNetDispenser(
     *     {
     *       authToken: 'your_auth_token',
     *       requestTimeout: 15,
     *     }
     * )
     *
     * @returns An instance of the TestNetDispenserApiClient class.
     */
    getTestNetDispenser(params = null) {
        return new TestNetDispenserApiClient(params);
    }
    /**
     * Returns a new `ApplicationClient` client, resolving the app by creator address and name.
     * @param details The details to resolve the app by creator address and name
     * @param cachedAppLookup A cached app lookup that matches a name to on-chain details; either this is needed or indexer is required to be passed in to this manager on construction.
     * @returns The `ApplicationClient`
     */
    getAppClientByCreatorAndName(details, cachedAppLookup) {
        return new ApplicationClient({ ...details, resolveBy: 'creatorAndName', findExistingUsing: cachedAppLookup ?? this.indexer }, this._algod);
    }
    /**
     * Returns a new `ApplicationClient` client, resolving the app by app ID.
     * @param details The details to resolve the app by ID
     * @returns The `ApplicationClient`
     */
    getAppClientById(details) {
        return new ApplicationClient({ ...details, resolveBy: 'id' }, this._algod);
    }
    /**
     * Returns a new typed client, resolving the app by creator address and name.
     * @param typedClient The typed client type to use
     * @param details The details to resolve the app by creator address and name
     * @param cachedAppLookup A cached app lookup that matches a name to on-chain details; either this is needed or indexer is required to be passed in to this manager on construction.
     * @returns The typed client instance
     */
    getTypedAppClientByCreatorAndName(typedClient, details, cachedAppLookup) {
        return new typedClient({ ...details, resolveBy: 'creatorAndName', findExistingUsing: cachedAppLookup ?? this.indexer }, this._algod);
    }
    /**
     * Returns a new typed client, resolving the app by app ID.
     * @param typedClient The typed client type to use
     * @param details The details to resolve the app by ID
     * @returns The typed client instance
     */
    getTypedAppClientById(typedClient, details) {
        return new typedClient({ ...details, resolveBy: 'id' }, this._algod);
    }
    /**
     * Retrieve client configurations from environment variables when defined or get defaults (expects to be called from a Node.js environment)
     *
     * If both `process.env.INDEXER_SERVER` and `process.env.ALGOD_SERVER` is defined it will use both along with optional `process.env.ALGOD_PORT`, `process.env.ALGOD_TOKEN`, `process.env.INDEXER_PORT` and `process.env.INDEXER_TOKEN`.
     *
     * If only `process.env.ALGOD_SERVER` is defined it will use this along with optional `process.env.ALGOD_PORT` and `process.env.ALGOD_TOKEN` and leave indexer as `undefined`.
     *
     * If only `process.env.INDEXER_SERVER` is defined it will use the default (LocalNet) configuration for both algod and indexer.
     *
     * It will return a KMD configuration that uses `process.env.KMD_PORT` (or port 4002) if `process.env.ALGOD_SERVER` is defined,
     * otherwise it will use the default LocalNet config unless it detects testnet or mainnet.
     * @example
     * ```typescript
     * const config = ClientManager.getConfigFromEnvironmentOrLocalNet()
     * ```
     * @returns The config for algod, indexer and kmd
     */
    static getConfigFromEnvironmentOrLocalNet() {
        if (!process || !process.env) {
            throw new Error('Attempt to get default client configuration from a non Node.js context; supply the config instead');
        }
        const [algodConfig, indexerConfig, kmdConfig] = process.env.ALGOD_SERVER
            ? [
                ClientManager.getAlgodConfigFromEnvironment(),
                process.env.INDEXER_SERVER ? ClientManager.getIndexerConfigFromEnvironment() : undefined,
                !process.env.ALGOD_SERVER.includes('mainnet') && !process.env.ALGOD_SERVER.includes('testnet')
                    ? { ...ClientManager.getAlgodConfigFromEnvironment(), port: process?.env?.KMD_PORT ?? '4002' }
                    : undefined,
            ]
            : [
                ClientManager.getDefaultLocalNetConfig('algod'),
                ClientManager.getDefaultLocalNetConfig('indexer'),
                ClientManager.getDefaultLocalNetConfig('kmd'),
            ];
        return {
            algodConfig,
            indexerConfig,
            kmdConfig,
        };
    }
    /** Retrieve the algod configuration from environment variables (expects to be called from a Node.js environment)
     *
     * Expects `process.env.ALGOD_SERVER` to be defined, and you can also specify `process.env.ALGOD_PORT` and `process.env.ALGOD_TOKEN`.
     */
    static getAlgodConfigFromEnvironment() {
        if (!process || !process.env) {
            throw new Error('Attempt to get default algod configuration from a non Node.js context; supply the config instead');
        }
        if (!process.env.ALGOD_SERVER) {
            throw new Error('Attempt to get default algod configuration without specifying ALGOD_SERVER in the environment variables');
        }
        return {
            server: process.env.ALGOD_SERVER,
            port: process.env.ALGOD_PORT,
            token: process.env.ALGOD_TOKEN,
        };
    }
    /**
     * Retrieve the indexer configuration from environment variables (expects to be called from a Node.js environment).
     *
     * Expects `process.env.INDEXER_SERVER` to be defined, and you can also specify `process.env.INDEXER_PORT` and `process.env.INDEXER_TOKEN`.
     */
    static getIndexerConfigFromEnvironment() {
        if (!process || !process.env) {
            throw new Error('Attempt to get default indexer configuration from a non Node.js context; supply the config instead');
        }
        if (!process.env.INDEXER_SERVER) {
            throw new Error('Attempt to get default indexer configuration without specifying INDEXER_SERVER in the environment variables');
        }
        return {
            server: process.env.INDEXER_SERVER,
            port: process.env.INDEXER_PORT,
            token: process.env.INDEXER_TOKEN,
        };
    }
    /** Returns the Algorand configuration to point to the free tier of the AlgoNode service.
     *
     * @param network Which network to connect to - TestNet or MainNet
     * @param config Which algod config to return - Algod or Indexer
     */
    static getAlgoNodeConfig(network, config) {
        return {
            server: `https://${network}-${config === 'algod' ? 'api' : 'idx'}.algonode.cloud/`,
            port: 443,
        };
    }
    /** Returns the Algorand configuration to point to the default LocalNet.
     *
     * @param configOrPort Which algod config to return - algod, kmd, or indexer OR a port number
     */
    static getDefaultLocalNetConfig(configOrPort) {
        return {
            server: `http://localhost`,
            port: configOrPort === 'algod' ? 4001 : configOrPort === 'indexer' ? 8980 : configOrPort === 'kmd' ? 4002 : configOrPort,
            token: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
        };
    }
    /**
     * Returns an algod SDK client that automatically retries on idempotent calls.
     *
     * @param config The config of the client
     * @example AlgoNode (testnet)
     * ```typescript
     *  const algod = ClientManager.getAlgodClient(ClientManager.getAlgoNodeConfig('testnet', 'algod'))
     *  await algod.healthCheck().do()
     * ```
     * @example AlgoNode (mainnet)
     * ```typescript
     *  const algod = ClientManager.getAlgodClient(ClientManager.getAlgoNodeConfig('mainnet', 'algod'))
     *  await algod.healthCheck().do()
     * ```
     * @example Custom (e.g. default LocalNet)
     * ```typescript
     *  const algod = ClientManager.getAlgodClient({server: 'http://localhost', port: '4001', token: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'})
     *  await algod.healthCheck().do()
     * ```
     */
    static getAlgodClient(config) {
        const { token, server, port } = config;
        const tokenHeader = typeof token === 'string' ? { 'X-Algo-API-Token': token } : token ?? {};
        const httpClientWithRetry = new AlgoHttpClientWithRetry(tokenHeader, server, port);
        return new algosdk.Algodv2(httpClientWithRetry, server);
    }
    /**
     * Returns an algod SDK client that automatically retries on idempotent calls loaded from environment variables (expects to be called from a Node.js environment).
     *
     * @example
     *  ```typescript
     *  // Uses process.env.ALGOD_SERVER, process.env.ALGOD_PORT and process.env.ALGOD_TOKEN
     *  const algod = ClientManager.getAlgodClientFromEnvironment()
     *  await algod.healthCheck().do()
     *  ```
     */
    static getAlgodClientFromEnvironment() {
        return ClientManager.getAlgodClient(ClientManager.getAlgodConfigFromEnvironment());
    }
    /**
     * Returns an indexer SDK client that automatically retries on idempotent calls
     *
     * @param config The config of the client
     * @param overrideIntDecoding Override the default int decoding for responses, uses MIXED by default to avoid lost precision for big integers
     * @example AlgoNode (testnet)
     * ```typescript
     *  const indexer = ClientManager.getIndexerClient(ClientManager.getAlgoNodeConfig('testnet', 'indexer'))
     *  await indexer.makeHealthCheck().do()
     * ```
     * @example AlgoNode (mainnet)
     * ```typescript
     *  const indexer = ClientManager.getIndexerClient(ClientManager.getAlgoNodeConfig('mainnet', 'indexer'))
     *  await indexer.makeHealthCheck().do()
     * ```
     * @example Custom (e.g. default LocalNet, although we recommend loading this into a .env and using the Default option instead)
     * ```typescript
     *  const indexer = ClientManager.getIndexerClient({server: 'http://localhost', port: '8980', token: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'})
     *  await indexer.makeHealthCheck().do()
     * ```
     * @example Override int decoding for responses
     * ```typescript
     *  const indexer = ClientManager.getIndexerClient(config, IntDecoding.BIGINT)
     * ```
     */
    static getIndexerClient(config, overrideIntDecoding) {
        const { token, server, port } = config;
        const tokenHeader = typeof token === 'string' ? { 'X-Indexer-API-Token': token } : token ?? {};
        const httpClientWithRetry = new AlgoHttpClientWithRetry(tokenHeader, server, port);
        const indexer = new Indexer(httpClientWithRetry);
        // Use mixed int decoding by default so bigints don't have lost precision
        indexer.setIntEncoding(overrideIntDecoding ?? IntDecoding.MIXED);
        return indexer;
    }
    /**
     * Returns an indexer SDK client that automatically retries on idempotent calls loaded from environment variables (expects to be called from a Node.js environment).
     *
     * @param overrideIntDecoding Override the default int decoding for responses, uses MIXED by default to avoid lost precision for big integers
     * @example
     *
     *  ```typescript
     *  // Uses process.env.INDEXER_SERVER, process.env.INDEXER_PORT and process.env.INDEXER_TOKEN
     *  const indexer = ClientManager.getIndexerClientFromEnvironment()
     *  await indexer.makeHealthCheck().do()
     *  ```
     */
    static getIndexerClientFromEnvironment(overrideIntDecoding) {
        return ClientManager.getIndexerClient(ClientManager.getIndexerConfigFromEnvironment(), overrideIntDecoding);
    }
    /**
     * Returns a KMD SDK client.
     *
     * KMD client allows you to export private keys, which is useful to (for instance) get the default account in a LocalNet network.
     *
     * @param config The config for the client
     * @example Custom (e.g. default LocalNet, although we recommend loading this into a .env and using the Default option instead)
     * ```typescript
     *  const kmd = ClientManager.getKmdClient({server: 'http://localhost', port: '4002', token: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'})
     * ```
     */
    static getKmdClient(config) {
        const { token, server, port } = config;
        return new Kmd(token, server, port);
    }
    /**
     * Returns a KMD SDK client that automatically retries on idempotent calls loaded from environment variables (expects to be called from a Node.js environment).
     *
     * @example
     *  ```typescript
     *  // Uses process.env.ALGOD_SERVER, process.env.KMD_PORT (or if not specified: port 4002) and process.env.ALGOD_TOKEN
     *  const kmd = ClientManager.getKmdClientFromEnvironment()
     *  ```
     */
    static getKmdClientFromEnvironment() {
        // We can only use Kmd on the LocalNet otherwise it's not exposed so this makes some assumptions
        // (e.g. same token and server as algod and port 4002 by default)
        return ClientManager.getKmdClient({ ...ClientManager.getAlgodConfigFromEnvironment(), port: process?.env?.KMD_PORT ?? '4002' });
    }
}

export { ClientManager };
