// @algorandfoundation/algokit-utils@6.2.1/types/indexer.js
'use strict';

/** Defines the what additional actions occur with the transaction https://developer.algorand.org/docs/rest-apis/indexer/#oncompletion */
exports.ApplicationOnComplete = void 0;
(function (ApplicationOnComplete) {
    ApplicationOnComplete["noop"] = "noop";
    ApplicationOnComplete["optin"] = "optin";
    ApplicationOnComplete["closeout"] = "closeout";
    ApplicationOnComplete["clear"] = "clear";
    ApplicationOnComplete["update"] = "update";
    ApplicationOnComplete["delete"] = "delete";
})(exports.ApplicationOnComplete || (exports.ApplicationOnComplete = {}));
/** Type of signature used by an account */
exports.SignatureType = void 0;
(function (SignatureType) {
    /** Normal signature */
    SignatureType["sig"] = "sig";
    /** Multisig */
    SignatureType["msig"] = "msig";
    /** Logic signature */
    SignatureType["lsig"] = "lsig";
})(exports.SignatureType || (exports.SignatureType = {}));
/** Delegation status of the account */
exports.AccountStatus = void 0;
(function (AccountStatus) {
    /** Indicates that the associated account is delegated */
    AccountStatus["Offline"] = "Offline";
    /** Indicates that the associated account used as part of the delegation pool */
    AccountStatus["Online"] = "Online";
    /** Indicates that the associated account is neither a delegator nor a delegate */
    AccountStatus["NotParticipating"] = "NotParticipating";
})(exports.AccountStatus || (exports.AccountStatus = {}));
