// @algorandfoundation/algokit-utils@6.2.1/types/logging.js
'use strict';

/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable no-console */
/** A logger implementation that writes to console */
const consoleLogger = {
    error: console.error,
    warn: console.warn,
    info: console.info,
    verbose: () => { },
    debug: console.debug,
};
const infoConsoleLogger = {
    error: console.error,
    warn: console.warn,
    info: console.info,
    verbose: () => { },
    debug: () => { },
};
const verboseConsoleLogger = {
    error: console.error,
    warn: console.warn,
    info: console.info,
    verbose: console.trace,
    debug: console.debug,
};
const warningConsoleLogger = {
    error: console.error,
    warn: console.warn,
    info: () => { },
    verbose: () => { },
    debug: () => { },
};
/** A logger implementation that does nothing */
const nullLogger = {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    error: function (message, ...optionalParams) { },
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    warn: function (message, ...optionalParams) { },
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    info: function (message, ...optionalParams) { },
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    verbose: function (message, ...optionalParams) { },
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    debug: function (message, ...optionalParams) { },
};

exports.consoleLogger = consoleLogger;
exports.infoConsoleLogger = infoConsoleLogger;
exports.nullLogger = nullLogger;
exports.verboseConsoleLogger = verboseConsoleLogger;
exports.warningConsoleLogger = warningConsoleLogger;
